pkg/controllers/provisioning/scheduling/modelinference.go

```diff
@@ -0,0 +1,181 @@
+		pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms = append(pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms, nodeSelectorTermsCopy...)
+	})
+
+	log.FromContext(ctx).
+		WithValues("Pod", klog.KObj(pod)).
+		V(1).Info(fmt.Sprintf("adding requirements derived from pod's inference flavors, %s", pretty.Concise(pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution)))
+
+	return nil
+}
+
+func (m *ModelInference) getInferenceFlavors(ctx context.Context, pod *v1.Pod) ([]llmazcoreapi.Flavor, error) {
+	modelName, ok := pod.Labels[llmazcoreapi.ModelNameLabelKey]
+	if !ok {
+		// Ignore the pod that is not created via llmaz's inference service.
+		return nil, nil
+	}
+
+	model := &llmazcoreapi.OpenModel{}
+	if err := m.kubeClient.Get(ctx, types.NamespacedName{Name: modelName}, model); err != nil {
+		return nil, fmt.Errorf("getting open model %q, %w", modelName, err)
+	}
+	modelFlavors := lo.FromPtrOr(model.Spec.InferenceConfig, llmazcoreapi.InferenceConfig{}).Flavors
+
+	serviceFlavorRawStr, ok := pod.Annotations[llmazinferenceapi.InferenceServiceFlavorsAnnoKey]
+	if !ok {
+		// Not all inference pods specify the inference service flavors.
+		return modelFlavors, nil
+	}
+
+	modelFlavorMap := lo.SliceToMap(modelFlavors, func(flavor llmazcoreapi.Flavor) (llmazcoreapi.FlavorName, llmazcoreapi.Flavor) {
+		return flavor.Name, flavor
+	})
+
+	var result []llmazcoreapi.Flavor
+	for _, flavorNameVal := range strings.Split(serviceFlavorRawStr, ",") {
+		flavor, ok := modelFlavorMap[llmazcoreapi.FlavorName(flavorNameVal)]
+		if !ok {
+			return nil, fmt.Errorf("unknown service inference flavor %q", flavorNameVal)
+		}
+		result = append(result, flavor)
+	}
+	return result, nil
+}
+
+func (m *ModelInference) ValidateInferenceFlavors(ctx context.Context, pod *v1.Pod) (err error) {
+	modelName, ok := pod.Labels[llmazcoreapi.ModelNameLabelKey]
+	if !ok {
+		// Ignore the pod that is not created via llmaz's inference service.
+		return nil
+	}
+
+	model := &llmazcoreapi.OpenModel{}
+	if err := m.kubeClient.Get(ctx, types.NamespacedName{Name: modelName}, model); err != nil {
+		return serrors.Wrap(fmt.Errorf("failed to validate open model, %w", err), "OpenModel", klog.KRef("", modelName))
+	}
+
+	serviceFlavorRawStr, ok := pod.Annotations[llmazinferenceapi.InferenceServiceFlavorsAnnoKey]
+	if !ok {
+		// Not all inference pods specify the inference service flavors.
+		return nil
+	}
+
+	// Get all flavors from the model and check if the service flavors are valid.
+	allFlavors := lo.SliceToMap(
+		lo.FromPtrOr(model.Spec.InferenceConfig, llmazcoreapi.InferenceConfig{}).Flavors,
+		func(flavor llmazcoreapi.Flavor) (llmazcoreapi.FlavorName, llmazcoreapi.Flavor) {
+			return flavor.Name, flavor
+		},
+	)
+	unknownFlavors := lo.Reject(strings.Split(serviceFlavorRawStr, ","), func(flavor string, _ int) bool {
```

Comment:
As mentioned, I think we can tolerate the error here. Generally this should be validated in creation in llmaz system.